OpenToken Library - token requests

quantumew opened the pull request:
OpenToken library with an interface for public and private token uploading and downloading. This also supports uploading from and downloading to files. This library currently does not follow self discovery links. That is on the TODO list.

kyle-long committed:
boilerplate configs

kyle-long committed:
initial commit of package.json

kyle-long committed:
He doesn't mind having his email on stuff

kyle-long committed:
initial commit of authorization builder

kyle-long committed:
placeholder

kyle-long committed:
to make it easier to use.

kyle-long committed:
to match our standard

quantumew committed:
Adding urijs and fixing typedef

quantumew committed:
Adding a logger.

quantumew committed:
Fixing module name and promisifying request.

quantumew committed:
Committing current progress.

quantumew committed:
Fixing some issues with the container to get it working.

quantumew committed:
Merge branch 'develop' into signed-requests

quantumew committed:
Wrong module name

quantumew committed:
Merge branch 'develop' into signed-requests

quantumew committed:
Actually fix it

quantumew committed:
Merge branch 'develop' into signed-requests

quantumew committed:
Fixing some minor issues.

quantumew committed:
Fixing logger

quantumew committed:
Merge branch 'develop' into signed-requests

kyle-long committed:
initial commit of test stuff

kyle-long committed:
Merge branch 'develop' of github.com:opentoken-io/opentoken-lib-js into develop

quantumew committed:
Fixing code and secret references in auth builder.

quantumew committed:
Merge branch 'develop' into signed-requests

kyle-long committed:
adding hasha and StreamConcat

kyle-long committed:
Merge branch 'develop' of github.com:opentoken-io/opentoken-lib-js into develop

quantumew committed:
Made some functionality changes.

Renamed create -> upload and handle streaming to a file.

kyle-long committed:
allowing body to be a stream

kyle-long committed:
Merge branch 'develop' into signed-requests

kyle-long committed:
passing along body

kyle-long committed:
Filling out main lib with one of our two main functions

kyle-long committed:
correcting typo

kyle-long committed:
handling async

kyle-long committed:
passing along loglevel

kyle-long committed:
not used

quantumew committed:
Add a sloppy initial test that is passing.

Broke the Date stuff into a date service. Fixed the test script.

kyle-long committed:
was not hashing with the secret

quantumew committed:
Added tests for upload and download to file. Also added helpful mocks.

These tests will be refactored once I get closer to finishing this work.

quantumew committed:
Fixing up uploadFromFile to work with a stream.

quantumew committed:
Added TokenRequest and put common functionality in token downloader.

Also update tests to work with new setup.

kyle-long committed:
initial commit of  tests

kyle-long committed:
Merge branch 'signed-requests' of github.com:opentoken-io/opentoken-lib-js into signed-requests

kyle-long committed:
Testing case insensitive headers

kyle-long committed:
reusing signature

kyle-long committed:
Asserting it logs stuff.

kyle-long committed:
we no longer use hasha

kyle-long committed:
fixing name

It will only ever be used as a javacript language

quantumew committed:
Add promisifier test and use container in signed-token-request

quantumew committed:
Removing broken test.

quantumew committed:
Update README with basic example and usage. Add factory for TokenRequest

quantumew committed:
Add better error handling with test.

Also add a opentoken lib test.

quantumew committed:
Fix tests and error response handling.

quantumew committed:
Fix uploadFromFile and complete tests.

There were other little fixes in here as well.

quantumew committed:
Fixing minor doc block and formatting.

lnagle commented on README.md:
```diff
@@ -1,2 +1,42 @@
+Usage
+-----
+
+The library exposes two functions `createSignedTokenRequests` and `createTokenRequest`. Let's go through an example of a common use case for token requests.
```
Add a colon after "functions".

lnagle commented on README.md:
```diff
@@ -1,2 +1,42 @@
+        // Public content can also be downloaded with the signedTokenRequests. Takes a stream or file name.
+        return tokenRequest.downloadToFile(token, "output.txt");
+    })
+
```
Need another line break here.

lnagle commented on README.md:
```diff
@@ -1,2 +1,42 @@
+SignedTokenRequest
+------------------
+
+`createSignedTokenRequests` returns an object containing two properties `public` and `private`. Each of these properties are instances of `SignedTokenRequest`, meant to be used with public and private endpoints. It exposes four functions `download(token)`, `downloadToFile(token, filePath)`, `upload(content)`, and `uploadToFile(filePath)`.
```
Add a colon after "properties" in the first sentence and one after "functions" in the third sentence.

lnagle commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+            this.body = body;
+            this.requestOptions = requestOptions;
+
+            // This is done here so that when I used it in two different places
```
I'm assuming "it" in this case is `this.headerNameList`, but it may be good to reword this so that an assumption isn't necessary.

lnagle commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+
+        /**
+         * Builds a string which are all the things that need to be
+         * hashed to build the signature.
```
The phrase "a string which are all the things" lacks subject/verb agreement.

lnagle commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+
+                if (headerName === "host") {
+                    // The Host header must be changed to lowercase because
+                    // DNS is case insensitive.
```
Do you mean to say "DNS is case sensitive"?

quantumew replied:
Kyle wrote this but no. DNS is case insensitive.

lnagle replied:
If it's case insensitive, then why does it need to be made lowercase?

lnagle replied:
This should be rephrased. Simply saying "The Host head is case sensitive" would do.

quantumew replied:
Finally I understand! Kyle added this comment because OpenToken specifically requires the host header to be lower case. I updated the comment to simply say that. f5731215919d3358b29714d6afdf3a4e87164f0d

lnagle commented on lib/container.js:
```diff
@@ -0,0 +1,66 @@
+    Log: "log",
+    parseLinkHeader: "parse-link-header",
+    request: "request",
+    Uri: "urijs",
```
Please bump this down a couple lines, for the sake of alphabetization.

lnagle commented on lib/container.js:
```diff
@@ -0,0 +1,66 @@
+    streamToPromise: "stream-to-promise"
+};
+
+// All modules that a factories (create an object).
```
Change "a" to "are".

lnagle commented on lib/content-hasher.js:
```diff
@@ -0,0 +1,104 @@
+
+    /**
+     * Responsible for hashing a list of content. The real reason to
+     * use this class over
```
Please complete this sentence.

lnagle commented on lib/request-handler.js:
```diff
@@ -0,0 +1,141 @@
+        var message;
+
+        // If OpenToken does not return a message in the body, I attempt to map
+        // the status code to an error. If all else fails I return an unknown error.
```
Add a comma after fails.

lnagle commented on lib/request-handler.js:
```diff
@@ -0,0 +1,141 @@
+            path = `${path}/${token}`;
+        }
+        if (isPublic) {
+            url.query({public: true});
```
Add newlines after the first curly brace and after true.

lnagle commented on lib/request-handler.js:
```diff
@@ -0,0 +1,141 @@
+            body = JSON.parse(body);
+        } catch (e) {
+            logger.debug("Failed to parse response body to JSON.");
+        }
```
Need a newline here, and in several other places in this file.

quantumew replied:
Saw this comment and wanted to respond. According to the style guide there should not be a newline here, unless I am missing something. https://github.com/tests-always-included/style-guide/blob/master/lang/js/index.md#indentation-and-whitespace

quantumew replied:
Ahh, so I didn't see the general section on newlines. I will fix this!

lnagle replied:
Resolved.

lnagle commented on lib/signed-token-request.js:
```diff
@@ -0,0 +1,161 @@
+ * @param {uri-js} Uri
+ * @return {SignedTokenRequest}
+ */
+module.exports = (bluebird, fs, host, logger, signer, StreamPassThrough, request, requestAsync, requestHandler, tokenDownloader, Uri) => {
```
This list should be alphabetized.

lnagle commented on lib/signed-token-request.js:
```diff
@@ -0,0 +1,161 @@
+
+            // Create a stream and when the original is read in
+            // the PassThrough stream is loaded with its data.
+            // The former used to calculate the body hash
```
Add "will be" between "former" and "used".

lnagle commented on package.json:
```diff
@@ -0,0 +1,54 @@
+{
+    "name": "opentoken-lib",
+    "version": "0.0.1",
```
The most recent I projects I've code reviewed have all started at 1.0.0, but I'm not sure that that's correct either. Perhaps something we should talk about as a team.

quantumew replied:
Yea, I think Kyle did the package but I don't really care either way. I am slightly disposed to starting from `1.0.0`. Just makes more sense. <major>.<minor>.<revision/or something> - fixed

lnagle replied:
Resolved.

lnagle commented on scripts/test:
```diff
@@ -0,0 +1,22 @@
+fi
+
+# Verify the coding standards are enforced.
+eslint lib/
```
Add `spec/` to this.

lnagle commented on scripts/test:
```diff
@@ -0,0 +1,22 @@
+    # is to use --dump.
+    echo "Executing codecov and generating coverage report."
+    codecov --dump > /dev/null
+fi
```
I've seen this implemented differently in Opentoken and something I reviewed for Kyle yesterday. Here's how it was done in those places: 

```
 if [[ "$CI" == "true" ]]; then
     codecov
 elif [[ -d ".git" ]]; then
     # Easiest way to generate coverage files and not upload the report
     # is to use --dump.
     echo "Executing codecov and generating coverage report."
     codecov --dump > /dev/null
 else
     # codecov runs git and hg(mercurial).  If the folder is not under source
     # control, codecov breaks with some nasty error.  This situation happens
     # during deployment, so we just don't worry about coverage during
     # deployment.
     echo "Not running code coverage report."
 fi
```

quantumew replied:
I am OK with our implementation. We use git and we aren't using the codecov badge or service at the moment.

lnagle replied:
Works for me.

lnagle commented on spec/lib/authorization-builder.spec.js:
```diff
@@ -0,0 +1,102 @@
+
+    beforeEach(() => {
+        loggerMock = require("../mock/logger-mock")();
+        dateMock = jasmine.createSpyObj("date", ["now"]);
```
`"now"` should be on it's own line, apart from either bracket.

lnagle commented on spec/lib/authorization-builder.spec.js:
```diff
@@ -0,0 +1,102 @@
+
+            // Important to note: The fact that the "signed-headers" are not
+            // cased the same as when they were used to build the signature
+            // is OK. So long as the ones were lowercase when used to build
```
Might consider changing "ones" to "signed-headers" for the sake of clarity.

quantumew replied:
That is what a pronoun is for right? So you don't have to keep saying the noun. I am going to leave it.

lnagle replied:
Fair enough.

fidian replied:
You should use the right pronoun.  "ones" is for individuals or the number.  Not as a reference for a group.  The right one would be "they" as you used in the previous line.

quantumew replied:
Indeed, skimmed it before. Fixed.

lnagle commented on spec/lib/authorization-builder.spec.js:
```diff
@@ -0,0 +1,102 @@
+
+            builder = new AuthorizationBuilder(code, secret, requestOptions, body);
+
+            // Important to note: The fact that the "signed-headers" are not
```
Does "signed-headers" need to be in quotes here?

quantumew replied:
Yes. I don't see a reason why not to?

lnagle replied:
Put another way, what is the reason for doing so? What do the quotes denote?

quantumew replied:
I like it. It denotes a particular property in the Authorization header. It is quoting a section of the header.

lnagle replied:
Fair enough.

lnagle commented on spec/lib/opentoken-lib.spec.js:
```diff
@@ -0,0 +1,43 @@
+    beforeEach(() => {
+        opentokenLib = require("../../lib/opentoken-lib")(host, "info");
+        signed = opentokenLib.createSignedTokenRequests(account.id, account.code, account.secret);
+        req = opentokenLib.createTokenRequest(account.id);
```
For readability, it would be preferable to have this inside of the it block that it relates to.

quantumew replied:
Not exactly sure what you are saying. I did it this way because of the possibility of adding multiple tests. I want to create this factory in the beforeEach. Perhaps I am not understanding

quantumew replied:
Like I said I think this is negligible so I changed it 9529d96a1af1e301509078bf3dea4dc262067b39 - The reason for doing it this way was the anticipation of multiple tests for `createTokenRequest`. There is only one test as of now.

lnagle commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,253 @@
+        container.register("request", requestMock);
+        responseMock = require("../mock/response-mock")();
+        container.register("logger", require("../mock/logger-mock")());
+        fsMock = jasmine.createSpyObj("fs", ["createWriteStream"]);
```
"createWriteStream" should be on its own line.

lnagle commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,253 @@
+        StreamReadable = container.resolve("StreamReadable");
+        url = `https://${host}/account/${account.id}/token`;
+        privateSigned = new SignedTokenRequest(account.id, false, account.code, account.secret);
+        publicSigned = new SignedTokenRequest(account.id, true, account.code, account.secret);
```
Can any of these spyOn or spy altering (`requestMock.get.andReturn`, for example) statements be moved into beforeEach blocks closer to where they are used? I'm assuming not all of them can without losing DRYness, but what can be moved would make any debugging done down the road easier. 

Also, moving anything that alters a spy or spies on a method to the bottom of this beforeEach would make debugging easier as well.

quantumew replied:
Moved them towards the bottom where possible. Don't want to move the get.andReturn because multiple tests utilize it. Also if I was debugging a tests I would look over beforeEach because that is where a lot of stuff gets sets up.

lnagle replied:
_thumbs up emoji_

lnagle commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,253 @@
+
+
+    /**
+     * Assert Authrization header.
```
Need an "o" in Authorization.

lnagle commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,253 @@
+    describe(".download", () => {
+        var method;
+
+        method = "GET";
```
Setting a variable here doesn't seem to provide much benefit or cut down on the number of characters written.

quantumew replied:
It is used in multiple places and thus I have a variable so if it changes there is one spot to change it. More of a personal preference thing.

lnagle replied:
Works for me.

fidian replied:
Doesn't work for me.  I only see `method` used in the first test.  I would expect "METHOD" to be defined inside the "successfully downloads' test or possibly within the resolved promise's callback, a mere 2 lines above where it is used.

quantumew replied:
Not worth the trouble. Just removing it.

lnagle commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,253 @@
+        it("successfuly uploads", () => {
+            var actualRequestOpts, contents, method, resp;
+
+            method = "POST";
```
See other comment about setting the method to a variable.

lnagle replied:
Resolved.

lnagle commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,253 @@
+            });
+
+            // Basically the same thing as above. Have to have this event fired after
+            // the signer has signed. I found no other way to do this then setTimeout.
```
Change "then" to "than".

lnagle commented on spec/lib/token-request.spec.js:
```diff
@@ -0,0 +1,61 @@
+    beforeEach(() => {
+        container = require("../../lib/container")(host);
+        bluebird = container.resolve("bluebird");
+        fsMock = jasmine.createSpyObj("fs", ["createWriteStream"]);
```
Drop "createWriteStream" onto it's own line.

lnagle93 commented on spec/lib/token-request.spec.js:
```diff
@@ -0,0 +1,61 @@
+            fsMock.createWriteStream.andReturn({
+                path: file
+            });
+            promise = tr.downloadToFile(token, file).then((actualFile) => {
```
Is there a reason you're setting this to a variable?

lnagle commented on spec/lib/token-request.spec.js:
```diff
@@ -0,0 +1,61 @@
+            requestMock.emit("response", responseMock);
+            mockedStream.emit("finish");
+
+            return promise;
```
Is there a reason you're returning something here?

quantumew replied:
Yes so the tests works. How else would I emit the events? The promise would never resolve

lnagle replied:
Haha works for me.

lnagle commented on spec/mock/logger-mock.js:
```diff
@@ -0,0 +1,18 @@
+        "warn"
+    ].forEach((methodName) => {
+        logger[methodName] = jasmine.createSpy(`logger.${methodName}`);
+    });
```
Alternate syntax:

```
logger = jasmine.createSpyObj("logger", [
    "console",
    "debug",
    "error",
    "info",
    "warn"
]);

return logger;
```

All of the methods will be created as spies.

quantumew replied:
Lol took this from OpenToken and didn't look at it but agreed

lnagle commented on spec/support/jasmine.json:
```diff
@@ -0,0 +1,3 @@
+{
```
You don't need this file. You call the helpers in your test script. 

A support/jasmine.json is only used in Jasmine versions 2 and up, I believe.

quantumew replied:
Yup, that is vestigal

lnagle commented on spec/mock/response-mock.js:
```diff
@@ -0,0 +1,60 @@
+                response.linkObjects.push(linkVal);
+            });
+        });
+    });
```
Add a line break here.

quantumew replied:
Spoke about this. Newline is not needed here.

quantumew committed:
Fix pull request comments.

Mostly grammar and style guide fixes.

quantumew commented:
@lnagle93 All code review fixes for pass one are here bd3896c50804025a8e96ae8faff5647cb09b650f

quantumew committed:
Moved place where object is instantiated in a test.

quantumew committed:
Update a comment

fidian commented:
Are you intending on squashing your commits?

fidian commented on README.md:
```diff
@@ -1,2 +1,43 @@
+        return signedTokenRequests.private.upload(token);
+    }).then((token) => {
+        // Grab public token from private endpoint.
+        return signedTokenRequests.private.download(token);
```
What is the difference between `signedTokenRequests.private.download()` and `signedTokenRequests.public.download()` if there is such a thing?  I ask because I'm assuming (perhaps incorrectly) that the `.public.` stuff is the same as `.private.`.

quantumew replied:
Here is the breakdown on that, `private` and `public` are instances of the same class. The only real difference is the property `isPublic` which is used in the query string for uploads. The downloads are the same. We saw no purpose in breaking this up further design wise because it would become more confusing.

fidian replied:
A request for a public token doesn't need to be signed.  Is downloading a public token without keys possible with the library?

quantumew replied:
Yes, using TokenRequest which is created with `opentokenLib.createTokenRequest(accountId)`. I am going to bring this up tomorrow at stand up to clarify our design decision on this.

fidian replied:
Works for me.  Resolved.

fidian commented on README.md:
```diff
@@ -1,2 +1,43 @@
+SignedTokenRequest
+------------------
+
+`createSignedTokenRequests` returns an object containing two properties: `public` and `private`. Each of these properties are instances of `SignedTokenRequest`, meant to be used with public and private endpoints. It exposes four functions: `download(token)`, `downloadToFile(token, filePath)`, `upload(content)`, and `uploadToFile(filePath)`.
```
It would be greatly beneficial to break out each of these methods as an API reference.

fidian commented on README.md:
```diff
@@ -1,2 +1,43 @@
+TokenRequest
+------------
+
+Provides an interface for unsigned token requests. `createTokenRequest` returns an instance of `TokenRequest`. It exposes two functions: `download(token)` and `downloadToFile(token, filePath)`.
```
Likewise, this would benefit from breaking out and documenting the functions separately.

quantumew replied:
Broke both of these out into API docs.

fidian commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+
+
+/**
+ * @typedef {Object} AuthorizationBuilder - Builds components for signing requests for OpenToken.
```
Hyphen should not exist.  Description starts immediately after the name you're defining.

quantumew replied:
Fixed

fidian commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+
+/**
+ * @typedef {Object} AuthorizationBuilder - Builds components for signing requests for OpenToken.
+ * @function buildHeaderValue
```
These are not documented right.  `@function` denotes the thing as a function in case the parser can't tell.  You've already denoted the thing as an Object on the previous line.

    @property {Function} buildHeaderValue

fidian commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+
+/**
+ * @param {ContentHasher} ContentHasher
+ * @param {log} logger
```
Nothing defines what `log` is or what properties you need.  I guess if that's supposed to be injected, we should define it like this:

    /**
     * @typedef {Object} log
     * @property {Function} debug
     * ... and so on ...
     */

However, since you are injecting this from your own logger, the typedef should be defined there.  You should use `@param {logger~Logger} logger` or similar here.

quantumew replied:
Fixed. It is from the log module so denoted as `log~Log`.

fidian commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+         * @param {string} code
+         * @param {string} secret
+         * @param {requestOptions} requestOptions
+         * @param {string|stream~Readable} body
```
When dealing with content, it is far more useful to use `Buffer` objects instead of strings.  Can your library be used with buffers?

Also, use parenthesis like so:

    * @param {(string|stream~Readable)} body

quantumew replied:
Agreed. Fixed.

fidian commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+            this.body = body;
+            this.requestOptions = requestOptions;
+
+            // This is done here so that when I used headerNameList in two
```
Change "used" to "use".

Good use of a comment!

quantumew replied:
Fixed

fidian commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+
+
+        /**
+         * Creates the authorization header value for the provided requestOptions.
```
requestOptions are not provided to this function.  Perhaps reword this comment and remove everything after "value"?

quantumew replied:
Reworded this. Btw a commit hash will follow at the end when I fix everything. I will also squash my commits when I merge.

fidian replied:
Commit at the end and the squash right before the merge is great, especially if I can review before the squash.

fidian commented on lib/authorization-builder.js:
```diff
@@ -0,0 +1,113 @@
+            hasher.push("\n");
+
+            if (this.body) {
+                hasher.push(this.body);
```
This will fail when dealing with strings of specific encodings, which is why I was hoping that you could use buffers earlier.

quantumew replied:
Indeed. Fixed

fidian commented on lib/container.js:
```diff
@@ -0,0 +1,66 @@
+
+Dizzy = require("dizzy");
+
+module.exports = (host, logLevel) => {
```
I think it would be wiser to have `logLevel` replaced with `options` and let people configure this via an `options` object.  It could have properties like `logLevel` or even be later expanded with `logger` to swap out the logger and other properties could enable/disable HTTPS-only mode or verification of SSL certificates.

quantumew replied:
Good idea. Fixed.

fidian commented on lib/container.js:
```diff
@@ -0,0 +1,66 @@
+    container.register("StreamReadable", require("stream").Readable).asValue();
+    container.register("StreamPassThrough", require("stream").PassThrough).asValue();
+
+    Object.keys(moduleDefs).forEach((moduleKey) => {
```
We can now use `registerBulk` if you want.

    container.registerBulk(moduleDefs).fromModule();
    container.registerBulk(factoryDefs).fromModule(__dirname).asFactory().cached();

quantumew replied:
Fixed

fidian commented on lib/content-hasher.js:
```diff
@@ -0,0 +1,106 @@
+
+
+/**
+ * @typedef {ContentHasher}
```
@typedef {Object} ContentHasher
    @property {Function} push

quantumew replied:
Fixed

fidian commented on lib/content-hasher.js:
```diff
@@ -0,0 +1,106 @@
+
+    /**
+     * Responsible for hashing a list of content. The real reason to
+     * use this class over a simple concated and hashed string is to
```
s/concated/concatenated/

quantumew replied:
Fixed.

fidian commented on lib/content-hasher.js:
```diff
@@ -0,0 +1,106 @@
+        /**
+         * Adds more content that will be used in the hash.
+         *
+         * @param {string|stream~Readable} content
```
Buffer?

quantumew replied:
This has been updated to `@param {(string|Buffer|stream~Readable)} content`. Headers and everything gets pushed to this so it takes all three.

fidian commented on lib/content-hasher.js:
```diff
@@ -0,0 +1,106 @@
+
+            concat = new StreamConcat(this.contentList);
+            hmac = crypto.createHmac(algorithm, this.secret);
+            hmac.setEncoding("hex");
```
Why isn't this controlled via a parameter?  It seems odd to allow flexibility in one area and no flexibility in the other.

I don't mind if you take out the algorithm entirely and hardcode SHA256 and hexadecimal or if you add an encoding parameter.  OpenToken currently can only support version 1 signatures, which must currently use SHA256 and hex only.

quantumew replied:
Good point. I changed it to allow the setting of encoding but I default it to hex.

fidian commented on lib/date-service.js:
```diff
@@ -0,0 +1,28 @@
+
+
+/**
+ * @typedef {dateService}
```
@typedef {Object} dateService
    @property {Function} now

quantumew replied:
Fixed

fidian commented on lib/date-service.js:
```diff
@@ -0,0 +1,28 @@
+
+
+/**
+ * Creates a dateService.
```
What is a `dateService`?  What does it do?

quantumew replied:
Added description

fidian commented on lib/logger.js:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Creates a logger.
+ *
+ * @param {log} Log
```
Should this be `@param {Log} log`?  It's confusing because you take `{log}` as input and return `{log}` as output.

Where is `{log}` defined as a typedef?

quantumew replied:
No typedef because it is from the log module. I fixed the type. `log~Log`

fidian commented on lib/opentoken-lib.js:
```diff
@@ -0,0 +1,63 @@
+
+
+/**
+ * @typedef opentokenLib
```
@typedef {Object} opentokenLib
    @property {Function} createSignedTokenRequests

quantumew replied:
Fixed

fidian commented on lib/opentoken-lib.js:
```diff
@@ -0,0 +1,63 @@
+
+
+    /**
+     * Creates signedTokenRequests.
```
This has got to be one of the very few instances where a plural name is appropriate.  I even started commenting on it and caught myself because you actually do create multiple things.

quantumew replied:
For sure. Sort of confusing name.

fidian commented on lib/opentoken-lib.js:
```diff
@@ -0,0 +1,63 @@
+
+    container = require("./container")(host, logLevel);
+    SignedTokenRequest = container.resolve("SignedTokenRequest");
+    TokenRequest = container.resolve("TokenRequest");
```
Instead of this:

    container = require("...")(...);
    X = container.resolve("x");
    function a() { ... }
    return { a }

How about this?

    container = require("...")(...);
    return container.call((X) => {
        function a() { ... }
        return { a }
    })

Would it simplify the number of calls to `container.resolve()` by using `container.call()` instead?

quantumew replied:
Hmm interesting. I think I prefer it the way it is. If I had more then two calls I would be on your side with this. I am open to changing it though if you have a strong preference to it.

fidian replied:
No preference, just showing an alternative that you may not have already known.

Resolved.

fidian commented on lib/promisifier.js:
```diff
@@ -0,0 +1,18 @@
+
+        return bluebird.promisifyAll(otherModule);
+    };
+};
```
What do you think about making a `dizzy-bluebird` module that does this for you?  I don't know if I like seeing the same code in three projects.

Perhaps not for merge request #1, but at least something we can drop in Jira and pick up if we have a spare hour?  The projects that need it would be DDQ, DDQ-backend-mysql, OpenToken and now OpenToken Lib JS.

quantumew replied:
Indeed. I am opening a task for it now.

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+
+/**
+ * @typedef {Object} requestHandler
+ * @function buildRequestOptions
```
@property {Function} buildRequestOptions

quantumew replied:
Fixed

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+
+
+/**
+ * @typedef {Object} requestHandler
```
I think that in general, these factories produce something more like `requestHandler~result` instead of `requestHandler` things.  Or perhaps `{opentoken-lib-js~requestHandler}` to avoid potential namespace collisions.  What do you think?

quantumew replied:
I thought about this myself  and forgot to mention it, before coming to a conclusion and changing all of my JSDoc statements we should discuss this in stand-up. I will bring it up.

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+ */
+module.exports = (bluebird, dateService, host, logger, Uri) => {
+    // This error object is a last ditch attempt to supply info if
+    // OpenToken does not provide a body.
```
I'm concerned with this particular comment.  If OpenToken does not provide a body, do you really want to fabricate a body?

quantumew replied:
This was poorly worded. This is simply for an error message if an error response is returned with no discernible message. Perhaps this functionality should just be removed?

fidian replied:
I think I am in favor of removing it if the code doesn't provide benefit to anyone using it.

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+     * @return {Object}
+     */
+    function buildRequestOptions(url, method, contentType) {
+        if (typeof contentType === "undefined") {
```
Do you care about only `undefined`?  I think that `!contentType` is a better check in this instance.

quantumew replied:
No, not sure why I did it this way. Fixed

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+                message = error.unknown;
+            }
+        }
+        message = `Error occurred: ${message}, HTTP status code: ${statusCode}.`;
```
Need newline to offset the `message =` line from the end of the `if` block.

quantumew replied:
Fixed

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        if (!message) {
+            if (Object.keys(error).indexOf(statusCode.toString()) >= 0) {
```
Whoa, that's ugly.  Here's an alternative.

    if (error[statusCode]) {

quantumew replied:
Haha yea, that is bad. That is a throw back to Python. I always forget I can do it that way. Fixed

fidian replied:
Was thinking about this.  I suppose my alternative would break if someone modified `Object.prototype.whatever = true` and the statuscode matched `whatever`, but both of our things would break.  The truly correct one would be

    if (error.hasOwnProperty(statusCode)) {

But that is not what I normally see.  Besides, how often do you see people modifying the `Object.prototype` to include properties that are numbers?  :-)

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+        statusCode = response.statusCode;
+        logger.debug(`Status Code: ${statusCode}`);
+        logger.debug(`Response Body: ${body}`);
+        if (statusCode > 299 || statusCode < 200) {
```
Newline above.

quantumew replied:
Fixed

fidian commented on lib/request-handler.js:
```diff
@@ -0,0 +1,148 @@
+     * Checks for an error response.
+     *
+     * @param {http~IncomingMessage} response
+     * @throw {Error}
```
This doesn't throw.  Perhaps other things that are called may throw, but this one does not.

quantumew replied:
Yea, this is vestigial. Removed it.

fidian commented on lib/signed-token-request.js:
```diff
@@ -0,0 +1,161 @@
+
+/**
+ * @typedef {Object} SignedTokenRequest
+ * @function download
```
@property {Function} download

quantumew replied:
Fixed

fidian commented on lib/signed-token-request.js:
```diff
@@ -0,0 +1,161 @@
+         * @param {string} token
+         * @return {Promise.<string>}
+         */
+        download(token) {
```
Style guide says that things returning promises should have `Async` as a suffix.  `downloadAsync()`, `uploadAsync()`.

quantumew replied:
All instances fixed

fidian commented on lib/signed-token-request.js:
```diff
@@ -0,0 +1,161 @@
+            url = requestHandler.createUrl(this.accountId, this.isPublic);
+            requestOptions = requestHandler.buildRequestOptions(url, "POST");
+
+            return signer.sign(this.code, this.secret, requestOptions, stream).then(() => {
```
I don't see how this works without eating up memory.  What if you're trying to upload 1GB of data?

quantumew replied:
A PassThrough stream is created which has content pushed to it when the original stream is  piped. Then we pass the stream to the signer which passes it to the AuthorizationBuilder which uses the ContentHasher to push signature content to it. The ContentHasher concatenates the streams and pipes them to the hash. We then use the PassThrough stream to stream data to OpenToken. This basically amounts to streaming the contents twice but I found no way around this. At no point should the entire contents of the stream be buffered in memory.

fidian commented on lib/signer.js:
```diff
@@ -0,0 +1,38 @@
+
+
+/**
+ * @typedef {signer}
```
@typedef {Object} signer
    @property {Function} sign

quantumew replied:
Fixed

fidian commented on lib/signer.js:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @return {Promise}
+     */
+    function sign(code, secret, requestOptions, body) {
```
Given how the rest of the code works, wouldn't you want to do something like

    // if string, convert string to a stream
    signerStream = new signer("stuff here")
    stream.pipe(signerStream).pipe(request)

That should work fine.  The current approach seems more confusing.

quantumew replied:
I am not sure how this would work with our current ContentHasher. Will have to look into this. Going to have to add a ticket for this as well. This will not be a trivial task. This streaming stuff was the bulk of the time so far so I can't imagine this approach will be much easier.

quantumew replied:
Discussed this. Will research this.

quantumew replied:
No more streaming uploads. Load the file in and use the buffer. Fixed

fidian commented on lib/token-downloader.js:
```diff
@@ -0,0 +1,72 @@
+
+/**
+ * @typedef {Object} tokenDownloader
+ * @function download
```
@property {Function} download

quantumew replied:
Fixed.

fidian commented on lib/token-downloader.js:
```diff
@@ -0,0 +1,72 @@
+ * @param {bluebird} bluebird
+ * @param {fs} fs
+ * @param {request} request
+ * @param {request} requestAsync
```
These two lines are not injecting the same thing.

Technically, you don't need request if you have requestAsync.

quantumew replied:
Fixed

fidian commented on lib/container.js:
```diff
@@ -0,0 +1,66 @@
+        container.register(moduleKey, factoryDefs[moduleKey]).fromModule(__dirname).asFactory().cached();
+    });
+    [
+        "fs",
```
fsAsync is not used.

quantumew replied:
Yup vestigial. Fixed

fidian commented on lib/token-downloader.js:
```diff
@@ -0,0 +1,72 @@
+            stream = fs.createWriteStream(stream);
+        }
+
+        return new Bluebird((resolve) => {
```
There's no chance of a rejected promise here, yet it is possible for the write stream to emit an error event when a file can't be created or when the disk is full.

quantumew replied:
Good point. Added an error handler that rejects the promise if it occurs. Also added a test case for this.

fidian commented on lib/token-request.js:
```diff
@@ -0,0 +1,73 @@
+
+/**
+ * @typedef {Object} SignedTokenRequest
+ * @function download
```
@property {Function} download

quantumew replied:
Fixed

fidian commented on package.json:
```diff
@@ -0,0 +1,54 @@
+    "description": "Used to interact with the OpenToken API.",
+    "main": "lib/opentoken-lib.js",
+    "scripts": {
+        "test": "./scripts/test"
```
It's nice to have the `watch` script as well during test development.  Did you want to add that?

quantumew replied:
Added

fidian commented on spec/lib/opentoken-lib.spec.js:
```diff
@@ -0,0 +1,43 @@
+            code: "code",
+            secret: "secret",
+            id: "accountId"
+        },
```
I am unsure how we proceed with `const`.  I think I like your formatting and I suggest you add it to the style guide.

quantumew replied:
Will do

fidian commented on spec/lib/opentoken-lib.spec.js:
```diff
@@ -0,0 +1,43 @@
+        });
+    });
+    describe(".createTokenRequest", () => {
+        it(".public", () => {
```
Grammatically this does not make sense.  The call to `it()` should describe the action that the object performs.

quantumew replied:
Fixed

fidian commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,256 @@
+        stream.push(content);
+        stream.push(null);
+
+        return stream;
```
This is the third time I saw this code.

Maybe you should make a Jasmine helper?

quantumew replied:
Broke this out into a helper.

fidian commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,256 @@
+        expect(authList[2].trim()).toEqual("signed-headers=content-type x-opentoken-date host");
+        expect(authList[3].trim()).toMatch("signature=[a-z0-9].*");
+    }
+    describe(".download", () => {
```
Newline before.

quantumew replied:
Fixed

fidian commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,256 @@
+
+            // This is horrible but I could not find a better solution to it.
+            // Have to wait for the signer.sign promise to resolve before emitting
+            // the appropriate events to complete the response.
```
Why not attach these `.emit()` calls to the promise returned to the `signer.sign()` promise?

    oldSign = signer.sign;
    signer.sign = () => {
        var args, returned;
        args = [].slice.call(arguments);
        returned = oldSign.apply(signer, arguments);
        returned.then(() => {
            requestMock.emit("response", responseMock);
            stream.emit("finish");
        });
        return returned;
    }

You could even do this in the initial test setup in a slightly different way.

    signer._sign = signer.sign
    signer.sign = () => {
        var args;
        args = [].slice.call(arguments);
        signer._signPromise = signer._sign.apply(signer, args);
        return signer._signPromise;
    }
    signer._signPromise = null;

    // then in your test you first call the code that calls signer.sign and then
    signer._signPromise.then(() => {
        emit("stuff");
    });

It may even be worth making a signer mock that does this sort of stuff automatically and simply returns a prescribed value.  The triggering of the promise could be externally controlled by the test or you could more easily hook into that promise.

quantumew replied:
Not needed anymore due to the removal of streaming.

fidian commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,256 @@
+
+            // Basically the same thing as above. Have to have this event fired after
+            // the signer has signed. I found no other way to do this than setTimeout.
+            setTimeout(() => {
```
Above solution would be great because you can leverage whatever promise that signer.sign returns and trigger further events.

fidian commented on spec/lib/signed-token-request.spec.js:
```diff
@@ -0,0 +1,256 @@
+                expect(signer.sign).toHaveBeenCalled();
+            });
+
+            // Same as above. Need to have a slight pause before emitting this.
```
:-(

quantumew commented on lib/signer.js:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @return {Promise}
+     */
+    function sign(code, secret, requestOptions, body) {
```
Also, just to clarify currently the ContentHasher takes a bunch of content whether it is a stream, a string, or Buffer, then when hash is called on it, it concatenates all of the content into one stream and pipes it. So unless I am missing something. This will be a big rework but we will see.

quantumew committed:
Code review fixes!

The biggest functionality change here is we no longer support streaming
uploads. We simply load the file into memory, sign the request and send
it in the body of the request.

Fix grammatical, JSDoc, and style guide errors.
Suffixed functions that return a promise with Async.
Added a watch script.

quantumew commented:
@fidian code review fixes 902d782e6f9a1e8e0d250c47d121c3fbd2522c9e - Going to give it one more once over and then you can make your second pass.

quantumew committed:
Fixing JSDoc blocks and update types in this module.

quantumew commented:
Second pass. Missed a few things 19a341c43663c257ced30df58026f1631c4b2821

quantumew committed:
Add true jasmine helper, update docs, and fix typo.

quantumew commented:
@fidian pass two fixed 45a521c6fb797f4715521c650b296c731dab6115

fidian commented:
Good for me, especially if you change `opentoken-lib-js` into `opentokenLibJs` globally.

quantumew committed:
Fix module name in JSDoc.

quantumew commented:
Fixed 86a0b771d3728b5813cc5f0ba90c0c0bbbdfedf2
